selfdrive/thermald/power_monitoring.py

```diff
@@ -59,92 +67,120 @@ def panda_current_to_actual_current(panda_current):
 
 
 class PowerMonitoring:
-  def __init__(self):
+  def __init__(self, params):
+    self.params = params
     self.last_measurement_time = None           # Used for integration delta
     self.power_used_uWh = 0                     # Integrated power usage in uWh since going into offroad
     self.next_pulsed_measurement_time = None
+    self.car_voltage_mV = 12e3                  # Low-passed version of health voltage
     self.integration_lock = threading.Lock()
 
+    car_battery_power_uWh = params.get("CarBatteryPower")
+    if car_battery_power_uWh is None:
+      # If unknown, we assume the car battery is almost dead
+      self.car_battery_power_uWh = (CAR_BATTERY_CAPACITY_uWh / 10)
+    else:
+      # Reset capacity if it's low
+      self.car_battery_power_uWh = max((CAR_BATTERY_CAPACITY_uWh / 10), int(car_battery_power_uWh))
+    
+
   # Calculation tick
   def calculate(self, health):
     try:
       now = sec_since_boot()
 
       # Check that time is valid
-      if datetime.datetime.fromtimestamp(now).year < 2019:
+      if datetime.datetime.now().year < 2019:
         return
 
-      # Only integrate when there is no ignition
       # If health is None, we're probably not in a car, so we don't care
-      if health is None or (health.health.ignitionLine or health.health.ignitionCan) or \
-         health.health.hwType == log.HealthData.HwType.unknown:
+      if health is None or health.health.hwType == log.HealthData.HwType.unknown:
         with self.integration_lock:
           self.last_measurement_time = None
           self.next_pulsed_measurement_time = None
           self.power_used_uWh = 0
         return
 
+      # Low-pass battery voltage
+      self.car_voltage_mV = ((health.health.voltage * CAR_VOLTAGE_LOW_PASS_K) + (self.car_voltage_mV * (1 -  CAR_VOLTAGE_LOW_PASS_K)))
+
+      # Cap the car battery power and save it in a param
+      self.car_battery_power_uWh = max(self.car_battery_power_uWh, 0)
+      self.car_battery_power_uWh = min(self.car_battery_power_uWh, CAR_BATTERY_CAPACITY_uWh)
+      self.params.put("CarBatteryPower", str(int(self.car_battery_power_uWh)))
```

Comment:
Let's decimate this to every second or two and make it a non blocking write (`put_nonblocking`)

Reply:
If you do a non blocking put, keep in mind the writes can happen in the wrong order. Is that fine for the rest of the implementation?

Reply:
Should be finem yes. The param is only used on init (after reboot for example). It's not critical that it's always the exact value, it's an approximation anyways